### PR TITLE
Adding api client

### PR DESCRIPTION
**Pull request by brettvniekerk**
An API client has been implemented consistently across all projects I've worked on or reviewed, I believe it would be beneficial to have it added to the react-template.

### CODE REVIEW

**Review by xiangzhe0812 — DISMISSED** [review]

**Review by xiangzhe0812 — DISMISSED** [review]

**Review by OptimusOpus — DISMISSED** [review]

**Review by bazmatic — CHANGES_REQUESTED** [review]
No need to force-typecast the Promise

File: src/apis/HTTPClient.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  protected get = <T>(
+    url: string,
+    axiosReqConfig?: AxiosRequestConfig
+  ): Promise<T> =>
+    this.client
+      .get(url, axiosReqConfig)
+      .then(this.onSuccess)
+      .catch(this.onError) as unknown as Promise<T>;
+
```
Comment (bazmatic) [inline src/apis/HTTPClient.ts:44]:
```suggestion
  protected get = <T>(
    url: string,
    axiosReqConfig?: AxiosRequestConfig
  ): Promise<T> => {
    return new Promise<T>((resolve, reject) => {
      this.client
        .get(url, axiosReqConfig)
        .then((response: AxiosResponse<T>) => {
          resolve(this.onSuccess(response));
        })
        .catch((error: AxiosError) => {
          reject(this.onError(error));
        });
    });
  };
```

File: src/apis/HTTPClient.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  protected get = <T>(
+    url: string,
+    axiosReqConfig?: AxiosRequestConfig
+  ): Promise<T> =>
+    this.client
+      .get(url, axiosReqConfig)
+      .then(this.onSuccess)
+      .catch(this.onError) as unknown as Promise<T>;
```
Comment (bazmatic) [inline src/apis/HTTPClient.ts]:
```suggestion
      .catch(this.onError);
```

File: src/apis/HTTPClient.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  protected post = <T>(
+    url: string,
+    data?: unknown,
+    reqConfig?: AxiosRequestConfig
+  ): Promise<T> =>
+    this.client
+      .post(url, data, reqConfig)
+      .then(this.onSuccess)
+      .catch(this.onError) as unknown as Promise<T>;
```
Comment (bazmatic) [inline src/apis/HTTPClient.ts]:
```suggestion
      .catch(this.onError);
```

File: src/apis/HTTPClient.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+  protected put = <T>(
+    url: string,
+    data?: unknown,
+    axiosReqConfig?: AxiosRequestConfig
+  ): Promise<T> =>
+    this.client
+      .put(url, data, axiosReqConfig)
+      .then(this.onSuccess)
+      .catch(this.onError) as unknown as Promise<T>;
```
Comment (bazmatic) [inline src/apis/HTTPClient.ts]:
```suggestion
      .catch(this.onError);
```

**Review by bitcoinbrisbane — DISMISSED** [review]

File: src/apis/Api.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import HTTPClient from "./HTTPClient";
+
+class Api extends HTTPClient {
+  public example = async () => "Example call";
+}
+
+const api = new Api(process.env.VITE_API_URL as string, true);
```
Comment (Preston-Harrison) [inline src/apis/Api.ts]: a null check for this API url would be good, instead of casting to string

**Review by xiangzhe0812 — DISMISSED** [review]

File: src/apis/HTTPClient.ts (hunk 2 above)
Comment (Preston-Harrison) [inline src/apis/HTTPClient.ts]: Instead of typecasting `as unknown as Promise<T>`, use `this.client.get<T>(url, ...etc`

**Review by Preston-Harrison — CHANGES_REQUESTED** [review]

**Review by Preston-Harrison — DISMISSED** [review]

File: src/apis/Api.ts (hunk 5 above)
Comment (brettvniekerk) [reply src/apis/Api.ts]: 👍

File: src/apis/HTTPClient.ts (hunk 2 above)
Comment (brettvniekerk) [reply src/apis/HTTPClient.ts]: Already done

**Review by Preston-Harrison — APPROVED** [review]

**Review by bazmatic — APPROVED** [review]

**Pull request merged**